### PR TITLE
fix: add Kubernetes port forward IPCs and events

### LINKED ISSUE #9480: Add port forwarding event and communication
### Is your enhancement related to a problem? Please describe

To maintain a store at UI level we need inter process communication and various event

### Describe the solution you'd like

Implement new functions for interprocess and events related to port forwarding

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jeffmaury**
Fixes #9480

### What does this PR do?

Add IPCs and events for Kubernetes port forwarding

### Screenshot / video of UI

N/A

### What issues does this PR fix or reference?

#9480 

### How to test this PR?

Run unit tests

- [x] Tests are covering the bug fix or the new feature

### CODE REVIEW

File: packages/main/src/plugin/kubernetes/kubernetes-client.ts [hunk 1]
```diff
@@ -1741,4 +1748,26 @@ export class KubernetesClient {
   public async refreshContextState(context: string): Promise<void> {
     return this.contextsState.refreshContextState(context);
   }
+
+  protected ensurePortForwardService(): KubernetesPortForwardService {
+    if (this.#portForwardService === undefined) {
+      this.#portForwardService = KubernetesClient.portForwardServiceProvider.getService(this);
```
Comment (axel7083) [inline packages/main/src/plugin/kubernetes/kubernetes-client.ts]:
```suggestion
      this.#portForwardService = KubernetesClient.portForwardServiceProvider.getService(this, this.apiSender);
```

File: packages/main/src/plugin/kubernetes/kubernetes-client.ts [hunk 2]
```diff
@@ -491,6 +498,8 @@ export class KubernetesClient {
     }
     this.setupKubeWatcher();
     this.apiResources.clear();
+    this.#portForwardService?.dispose();
```
Comment (axel7083) [inline packages/main/src/plugin/kubernetes/kubernetes-client.ts:501]: Why do we dispose in the constructor ?

File: packages/main/src/plugin/kubernetes/kubernetes-port-forward-service.ts [hunk 3]
```diff
@@ -73,7 +75,10 @@ export class KubernetesPortForwardService implements IDisposable {
   constructor(
     private configManagementService: ConfigManagementService,
     private forwardingConnectionService: PortForwardConnectionService,
-  ) {}
+    private apiSender: ApiSenderType,
+  ) {
+    this.apiSender.send('kubernetes-port-forwards-update');
```
Comment (axel7083) [inline packages/main/src/plugin/kubernetes/kubernetes-port-forward-service.ts]: Not needed in the constructor ?

File: packages/main/src/plugin/kubernetes/kubernetes-port-forward-service.ts [hunk 4]
```diff
@@ -86,7 +91,9 @@ export class KubernetesPortForwardService implements IDisposable {
    * @see UserForwardConfig
    */
   async createForward(config: UserForwardConfig): Promise<ForwardConfig> {
-    return this.configManagementService.createForward(config);
+    const forwardConfig = await this.configManagementService.createForward(config);
+    this.apiSender.send('kubernetes-port-forwards-update');
```
Comment (axel7083) [inline packages/main/src/plugin/kubernetes/kubernetes-port-forward-service.ts]: To have similar update pattern than the kubernetes resources, we should provide the array in the event 
```suggestion
    this.apiSender.send('kubernetes-port-forwards-update', this.listForwards());
```

File: packages/main/src/plugin/kubernetes/kubernetes-port-forward-service.ts [hunk 5]
```diff
@@ -96,7 +103,8 @@ export class KubernetesPortForwardService implements IDisposable {
    * @see UserForwardConfig
    */
   async deleteForward(config: UserForwardConfig): Promise<void> {
-    return this.configManagementService.deleteForward(config);
+    await this.configManagementService.deleteForward(config);
+    this.apiSender.send('kubernetes-port-forwards-update');
```
Comment (axel7083) [inline packages/main/src/plugin/kubernetes/kubernetes-port-forward-service.ts]:
```suggestion
    this.apiSender.send('kubernetes-port-forwards-update', this.listForwards());
```

File: packages/main/src/plugin/kubernetes/kubernetes-port-forward-service.ts (hunk 3 above)
Comment (jeffmaury) [reply packages/main/src/plugin/kubernetes/kubernetes-port-forward-service.ts]: I added it there in order to clear the store when there is a context switch (thus a new service is created)

File: packages/main/src/plugin/kubernetes/kubernetes-client.ts [hunk 6]
```diff
@@ -1741,4 +1750,26 @@ export class KubernetesClient {
   public async refreshContextState(context: string): Promise<void> {
     return this.contextsState.refreshContextState(context);
   }
+
+  protected ensurePortForwardService(): KubernetesPortForwardService {
+    if (this.#portForwardService === undefined) {
+      this.#portForwardService = KubernetesClient.portForwardServiceProvider.getService(this, this.apiSender);
+    }
```
Comment (benoitf) [inline packages/main/src/plugin/kubernetes/kubernetes-client.ts]:
```suggestion
    if (!this.#portForwardService) {
      this.#portForwardService = KubernetesClient.portForwardServiceProvider.getService(this, this.apiSender);
    }
```
nitpick

File: packages/main/src/plugin/kubernetes/kubernetes-port-forward-service.ts (hunk 3 above)
Comment (axel7083) [inline packages/main/src/plugin/kubernetes/kubernetes-port-forward-service.ts]:
```suggestion
    this.apiSender.send('kubernetes-port-forwards-update', this.listForwards());
```

Comment (axel7083) [reply packages/main/src/plugin/kubernetes/kubernetes-port-forward-service.ts]: Noted!

**Review by benoitf — APPROVED** [review]

**Review by axel7083 — APPROVED** [review]
LGTM

**Review by axel7083 — CHANGES_REQUESTED** [review]
So sorry, but after testing it, I provided the wrong solution:

I recommended to add the objects when sending the events as followed

```
this.apiSender.send('kubernetes-port-forwards-update', this.listForwards());
```

This is a problem because `listForwards` is async! (why?)

IMO it should clearly not be async... but this is because the interface ForwardConfigStorage declare it async
 
https://github.com/containers/podman-desktop/blob/234674a252f1fef3051e72c6fc61f7cb24ff4e66/packages/main/src/plugin/kubernetes/kubernetes-port-forward-storage.ts#L58 

We cannot just remove the promise, because we have an implementation using the async (`FileBasedConfigStorage`), 

Therefore we need to do the following

```diff
- this.apiSender.send('kubernetes-port-forwards-update', this.listForwards());
+ this.apiSender.send('kubernetes-port-forwards-update', await this.listForwards());
```

**Review by axel7083 — APPROVED** [review]
LGTM !

**Pull request merged**